proposals/0540-jsem.rst

```diff
@@ -0,0 +1,384 @@
+two salient points:
+
+A. There is a hidden cost in having a GHC program (such as ``ghc`` itself) run on
+   fewer CPU cores than its capabilities: the stop-the-world cost of garbage
+   collection becomes much more expensive. If we were to give GHC a fixed number
+   ``N`` of capabilities, then parallel garbage collections in that GHC would
+   recruit ``N`` OS threads. Instead we propose that GHC will adjust its number
+   of capabilities, via ``setNumCapabilities`` [8]_ , to the number of modules
+   it is compiling.
+   Note that GHC already calls ``setNumCapabilities`` with the argument to ``-j``.
+
+   We have concerns that rapidly adjusting the number of capabilities may have
+   adverse effects: we propose that GHC will limit the rate at which it can
+   release tokens to the semaphore so as to reduce this thrashing.
+   The exact mechanism of rate limiting is unspecified here.
+
+B. We believe that GHC invocations should not be too eager to release
+   semaphore tokens. This skews the balance in favour of in-unit parallelism
+   (1 unit with many capabilities), and against compiling many units in parallel
+   (many units with a single capability).
+   The justification is that the memory used by compiling units can be released
+   before starting another parallel process. Were GHC to release semaphore tokens
+   too eagerly, it could end up compiling a large number of packages in parallel
+   which each have a large loaded EPS. Combined, this will use a significant amount
+   of memory.
+
+   To achieve this, we propose a minimum delay between acquiring a resource and
+   releasing a resource. This allows us to prioritise completing a single large
+   package before moving on to other work. The exact mechanism through which this
+   is achieved is left unspecified in this proposal.
+
+Examples
+--------
+
+Let us explain how we envision ``cabal`` handle the following build plan, with
+8 capabilities.
+
+.. image:: jsem_modules_plain.svg
+  :width: 600
+  :alt: Same dependency graph as before: a big unit at the top and bottom,
+        and many single-module units in the middle.
+
+1. To start, ``cabal`` would create a semaphore ``⟨sem⟩`` with 8 available tokens.
+
+2. Next, we compile the ``Bot`` unit, which is a large unit, with many modules,
+   which sits at the bottom of the dependency graph and must thus be compiled
+   before anything else.
+
+   a. ``cabal`` acquires one token from the semaphore and spawns one
+      ``ghc --make -jsem ⟨sem⟩`` invocation.
+   b. This invocation of ``ghc`` notices it has a lot of work to do (many modules
+      to compile from the ``Bot`` unit), so it requests more resources from the
+      semaphore. As no other processes are competing for semaphore tokens, this
+      GHC invocation obtains the remaining 7 tokens.
+   c. ``ghc`` finishes compiling the ``Bot`` unit, releasing the 7 tokens it
+      acquired.
+   d. ``cabal`` notices the ``ghc`` subprocess has terminated, and releases
```

Comment:
What happens if GHC dies suddenly, before having released the additional 7 tokens? Does Cabal somehow force those to be released as well? I'm concerned about possible resource starvation issues.

Reply:
If GHC dies suddenly then it will indeed not return the 7 tokens it has taken, but cabal can still return the one it owns so things can proceed for a bit longer with less parallelism. However, if GHC dies like this then the compilation has failed so cabal should find this out and report the failure to the user promptly.

Reply:
> However, if GHC dies like this then the compilation has failed so cabal should find this out and report the failure to the user promptly

True, but we could rephrase the scenario a bit such that GHC fails to return the additional tokens, but does finish compilation successfully so Cabal continues with seriously limited parallelism.

What I'm curious about here is whether there's any benefit in establishing a parent-child relationship between tokens. Imagine that in step (d) when Cabal returns the implicit token, any tokens that GHC had requested are automatically returned as well, even if GHC failed to explicitly return them.

Is that overcomplicating things?

Reply:
The problem with your suggestion is that tokens are directly requested from the semaphore, so Cabal doesn't know who owns which tokens, and thus can't know how many tokens to release. This is by design: coordination is achieved via the semaphore only.